xapian-letor/include/xapian-letor/ranker.h

```diff
@@ -193,7 +195,7 @@ class XAPIAN_VISIBILITY_DEFAULT Ranker : public Xapian::Internal::intrusive_base
   protected:
     /// Method to train the model. Overridden in ranker subclass.
     virtual void
-    train(const std::vector<Xapian::FeatureVector> & training_data) = 0;
+    train(const vector<vector<Xapian::FeatureVector>> & training_data) = 0;
```

Comment:
This (and all the related changes) is under-described in the commit message. Before, we were dealing with a vector of `FeatureVector` objects, ie one `FeatureVector` per document in the input MSet. I _think_ (by reading the changes) that we now have this once per query in the training set?

If this is what is going on, then we can choose better variable names in the training code, write some slightly better docstrings, and probably also more generously comment the implementations. For algorithm implementations, it can be helpful to explain quite a lot of the code, linking back to specific bits of the paper by section or formula if that helps make it clear what's going on. (The fact that you've had to fix the rankers in this way suggests we could really do with more clarity around what the code is doing and how that is a correct implementation!)